builtin/logical/transit/lock_manager.go

```diff
@@ -0,0 +1,337 @@
+	// to release the lock mutex right away to avoid the possibility of
+	// deadlock by trying to grab the second lock
+
+	// Check to make sure it hasn't been created since
+	lock = lm.locks[name]
+	if lock != nil {
+		lm.lockMutex.Unlock()
+		if writeLock {
+			lock.Lock()
+		} else {
+			lock.RLock()
+		}
+		return
+	}
+
+	lock = &sync.RWMutex{}
+	lm.locks[name] = lock
+	lm.lockMutex.Unlock()
+	if writeLock {
+		lock.Lock()
+	} else {
+		lock.RLock()
+	}
+}
+
+func (lm *lockManager) UnlockPolicy(name string, writeLock bool) {
+	lm.lockMutex.RLock()
+	lock := lm.locks[name]
+	lm.lockMutex.RUnlock()
+
+	if writeLock {
+		lock.Unlock()
+	} else {
+		lock.RUnlock()
+	}
+}
+
+func (lm *lockManager) GetPolicy(storage logical.Storage, name string) (*Policy, bool, error) {
+	p, lt, _, err := lm.getPolicyCommon(storage, name, false, false)
+	return p, lt, err
+}
+
+func (lm *lockManager) GetPolicyUpsert(storage logical.Storage, name string, derived bool) (*Policy, bool, bool, error) {
+	return lm.getPolicyCommon(storage, name, true, derived)
+}
+
+// When the function returns, a lock will be held on the policy if err == nil.
+// The type of lock will be indicated by the return value. It is the caller's
+// responsibility to unlock.
+func (lm *lockManager) getPolicyCommon(storage logical.Storage, name string, upsert, derived bool) (p *Policy, lockType bool, upserted bool, err error) {
+	// If we are using a cache, lock it now to avoid having to do really
+	// complicated lock juggling as we call various functions. We'll also defer
+	// the store into the cache.
+	lockType = shared
+	lm.LockPolicy(name, shared)
+
+	if lm.CacheActive() {
+		lm.globalMutex.RLock()
+		p = lm.cache[name]
+		if p != nil {
+			defer lm.globalMutex.RUnlock()
+			return
+		}
+		lm.globalMutex.RUnlock()
+
+		// When we return, since we didn't have the policy in the cache, if
+		// there was no error, write the value in.
+		defer func() {
+			if err == nil {
+				lm.globalMutex.Lock()
+				defer lm.globalMutex.Unlock()
+				// Make sure a policy didn't appear
+				exp := lm.cache[name]
+				if exp != nil {
+					p = exp
+					return
+				}
+
+				lm.cache[name] = p
+			}
+		}()
+	}
+
+	p, err = lm.getStoredPolicy(storage, name)
+	if err != nil {
+		defer lm.UnlockPolicy(name, shared)
+		return
+	}
+
+	if p == nil {
+		if !upsert {
+			defer lm.UnlockPolicy(name, shared)
+			return
+		}
+
+		// Get an exlusive lock; on success, check again to ensure that no
+		// policy exists. Note that if we are using a cache we will already be
+		// serializing this entire code path and it's currently the only one
+		// that generates policies, so we don't need to check the cache here;
+		// simply checking the disk again is sufficient.
+		lm.UnlockPolicy(name, shared)
+		lockType = exclusive
+		lm.LockPolicy(name, exclusive)
+
+		p, err = lm.getStoredPolicy(storage, name)
+		if err != nil {
+			defer lm.UnlockPolicy(name, exclusive)
+			return
```

Comment:
Can we remove the naked returns?

https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters